### DISCUSSION TITLE
CPU Thread Limit

### DISCUSSION THREAD

**Discussion by WillianJunior** [halide/Halide#5487] [Q&A] [answered]
Hi,

I need to limit the maximum number of CPU cores used by a Halide pipeline. From what I searched this should be done with `halide_set_num_threads(int)`. However, when I just call it on my code it won't compile with an "undefined reference to `halide_set_num_threads'". How can I use this function properly?

Thanks in advance for the help

**Comment by abadams** [comment, marked as answer]
That function should work when using Halide AOT-compiled. If you're using Halide as a JIT compiler, set the environment variable HL_NUM_THREADS to the max number of threads you want to use.

**Reply by WillianJunior** [reply]
I'm using JIT. Sorry for asking, but how does one set this environment variable? You mean at compilation of the source code with the JIT portion?

**Reply by WillianJunior** [reply]
Found it. For future references, use stdlib.h setenv()
https://pubs.opengroup.org/onlinepubs/9699919799/functions/setenv.html

**Reply by abadams** [reply]
It should be set when the realize call takes place. E.g. setenv("HL_NUM_THREADS", "16", 1) just before the first call to realize() to tell it to use 16 threads.

**Reply by WillianJunior** [reply]
Thanks again for the help. Just a curiosity, why was it decided that the CPU max thread count for JIT compilation should be set through environment variables instead of having a Halide::Runtime function set it, similar to AOT?

**Reply by abadams** [reply]
It's not an intentional design. Most production use of Halide is AOT-compiled, so the API to control the runtime is fairly complete there. In the JIT case the runtime API is not fully plumbed through.
